python/pyarrow/filesystem.py

```diff
@@ -371,9 +371,99 @@ def walk(self, path, refresh=False):
                 yield tup
 
 
-def _sanitize_s3(path):
+class AzureDatalakeFileSystemWrapper(DaskFileSystem):
+    """
+    Wraps the Gen1 Azure Datalake Filesystem client.
+    """
+
+    @implements(FileSystem.isdir)
+    def isdir(self, path):
+        path = _sanitize_remote_path(_stringify_path(path))
+        try:
+            contents = self.fs.ls(path)
+            if len(contents) == 1 and contents[0] == path:
+                return False
+            else:
+                if not any(ftype in path for ftype in
+                           ['parquet', 'parq', 'metadata']):
+                    raise ValueError('Directory is not a partition of'
+                                     ' *.parquet. Try passing a globstring.')
+                return True
+        except OSError:
+            return False
+
+    @implements(FileSystem.isfile)
+    def isfile(self, path):
+        path = _sanitize_remote_path(_stringify_path(path))
+        try:
+            contents = self.fs.ls(path)
+            return len(contents) == 1 and contents[0] == path
+        except OSError:
+            return False
+
+    def walk(self, path, invalidate_cache=True, refresh=False):
+        """
+        Directory tree generator, like os.walk
+        Generator version of what is in ADLFSClient, which yields a
+        flattened list of files
+        """
+        path = _sanitize_remote_path(_stringify_path(path))
+        directories = set()
+        files = set()
+
+        for key in list(self.fs._ls(path, invalidate_cache=invalidate_cache)):
+            path = key['name']
+            if key['type'] == 'DIRECTORY':
+                directories.add(path)
+            elif key['type'] == 'FILE':
+                pass
+            else:
+                files.add(path)
+
+        files = sorted([posixpath.split(f)[1] for f in files
+                        if f not in directories])
+        directories = sorted([posixpath.split(x)[1]
+                              for x in directories])
+
+        yield path, directories, files
+
+        for directory in directories:
+            for tup in self.walk(directory, refresh=refresh):
+                yield tup
+
+
+def _sanitize_remote_path(path):
```

Comment:
Hmm, we should instead have a specific function for each filesystem implementation. So, beside `_sanitize_s3`, add `_sanitize_adl`. Also, can you add unit tests for this in `test_filesystem.py`?